<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <link href='https://fonts.googleapis.com/css?family=Chivo:900' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen" />
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen" />
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print" />
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <title>ARC by IBEST</title>
  </head>
<!--
<script type="text/javascript">
function Toggle(tt) {
    var el = document.getElementById(tt);
    if (el.style.display == "block") {
        el.style.display = "none";
    }
    else {
        el.style.display = "block";
    }
}
</script>
-->
  <body>
    <div id="container">
      <div class="inner">

        <header>
          <h1>ARC</h1>
          <h2>Assembly by Reduced Complexity</h2>
        <img src="./images/Hickey_arc-float.png", width="300", height='280', style="float:left; margin:0px 0px 0px 10px" />
        <img src="./images/Pennell_arc_logo2.png", width="300", height='150' style="float:right; margin:0px 0px 0px 10px" />

        </header>
        <br><br><br><br><br><br><br><br><br><br><br><br><br>


        <center>Which logo do you prefer?</center>

        <br><br>

        <section id="downloads" class="clearfix">
          <a href="https://github.com/ibest/ARC/zipball/master" id="download-zip" class="button"><span>Download .zip</span></a>
          <a href="https://github.com/ibest/ARC/tarball/master" id="download-tar-gz" class="button"><span>Download .tar.gz</span></a>
          <a href="https://github.com/ibest/ARC" id="view-on-github" class="button"><span>View on GitHub</span></a>
        </section>

        <hr>

        <section id="main_content">
          <h3>Welcome to the Assembly by Reduced Complexity application page.</h3>

          <p>ARC is a pipeline which facilitates iterative, reference guided <em>de novo</em> assemblies with the intent of:
          <ol>
          <li>Reducing time in analysis and increasing accuracy of results by only considering those reads which should assemble together.</li>
          <li>Reducing/removing reference bias as compared to mapping based approaches. </li>
          </ol>
          The software is designed to work in situations where a whole-genome assembly is not the objective, but rather when the researcher wishes to assemble discreet 'targets' contained within next-generation shotgun sequence data. ARC decomplexifies the traditionally difficult problem of assembly by breaking the reads into small, manageable subsets which can then be assembled quickly and efficiently in parallel. Applications include those in which the researcher wishes to <em>de novo</em> assemble specific content and a set of semi-similar reference targets is available to initialize the assembly process.</p>


          <h4>Important Links</h4>
          <ul>
          <li>Please post questions, success stories, and topics for discussion here:<a href="https://groups.google.com/forum/#!forum/arc-assembly">Google Group</a></li>
          <li>Bug reports and suggested features <a href="https://github.com/ibest/ARC/issues">GitHub Issues</a></li>
          </ul>

          <div id="TOC">
          <h4>Table of Contents</h4>
          <ul>
          <li><a href="#ARCUseful">ARC Uses </a></li>
          <li><a href="#DesignObjectives">ARC Design Objectives </a></li>
          <li><a href="#ARCAlgorithm">ARC Algorithm</a></li>
          <li><a href="#ARCMotivation">ARC Motivation</a></li>
          <li><a href="#ARCInstallation">Installing ARC</a></li>
          <li><a href="#PatchingBLAT">Patching BLAT for Fastq support</a></li>
          <li><a href="#ARCTesting">Testing ARC install</a></li>
          <li><a href="#ARCUsage">How to use ARC</a></li>
          <li><a href="#ARCOutputs">ARC output files</a></li>
          <li><a href="#ARCConfig">Configuring ARC</a></li>
          <li><a href="#TipsAndTricks">Tips and tricks</a></li>
          <li><a href="#Contact">Contact us</a></li>

          </ul>
          </div>

          <!-- <a href="javascript:Toggle('ARCUseful');">  -->
          <div id="ARCUseful">
          <h4>ARC may be useful for:</h4>
          </a>
          <ul>
              <li>Assembly of mitochondrial genomes using a distantly related reference</li>
              <li>Assembly of exon capture data</li>
              <li>Assembly of bacterial plasmids</li>
              <li>Assembly of viral genomes</li>
              <li>Assembly of chloroplast genomes</li>
              <li>Transcriptomes (with plans for improved support in the future)</li>
              <h7><a href="#TOC">Table of contents</a></h7>
          </ul>
          </div>


          <div id="DesignObjectives">
          <h4>ARC design objects:</h4>
          <ul>
              <li>Break large, complex problems into smaller manageable chunks.</li>
              <li>Reduce memory footprint requirements (many assemblies should work on a desktop/labtop PC).</li>
              <li>Be highly scalable, running multiple jobs simultaneously in parallel.</li>
              <li>Be easy to use, portable and simple to configure.</li>
              <li>Be extensible with support for add on scripts and post-processing features.</li>
              <h7><a href="#TOC">Table of contents</a></h7>
          </ul>
          </div>

          <div id="ARCAlgorithm">
          <h4>The algorithm:</h4>

          <img src="./images/ARC2.png", width="267", height="188", style="float:right; margin:0px 0px 0px 10px" />
           <ol>
              <li>Map reads against a set of targets using BLAT or Bowtie2</li>
              <li>Extract mapped reads</li>
              <li>Assemble mapped reads into contigs using Roche/Newbler or Spades assemblers</li>
              <li>Map reads against the newly formed contigs</li>
              <li>Iterate until stopping conditions have been met</li>
              <h7><a href="#TOC">Table of contents</a></h7>
          </ol>
          </div>


<!-- <p><a href="javascript:Toggle();">Toggle me...</a></p>

        <a href="javascript:Toggle();"><h4>Motivation behind ARC</h4></a>
        <div id="ToggleTarget">
-->
        <div id="ARCMotivation">
        <h4> ARC Motivation </h4>
        <h5> ARC was designed to address the following problems: </h5>
        <ul>
        <li>In many experiments, <em>de novo</em> assembly of the full dataset is slow, resource intensive, and the end result is difficult to analyze because thousands of contigs are produced. Furthermore, it is difficult to take advantage of additional information available from previously assembled, but distantly related sequences during the assembly stage.</li>

        <li>Mapping based approaches also have limitations due to regions of low sequence identity where reads cannot be mapped as described by Heng Li (emphasis added):
        <blockquote cite="http://bioinformatics.oxfordjournals.org/content/early/2012/05/07/bioinformatics.bts280.abstract">"At present, the standard approach to variant calling is to map raw sequence reads against a reference genome and then to detect differences from the reference. It is well established and has been proved to work from a single sample to thousands of samples (1000 Genomes Project Consortium, 2010). Nonetheless, a fundamental flaw in this mapping based approach is that mapping algorithms ignore the <strong>correlation between sequence reads</strong>. They are unable to take full advantage of data and may produce inconsistent outputs which complicate variant calling. This flaw has gradually attracted the attention of various research groups who subsequently proposed several methods to alleviate the effect, including post alignment filtering (Ossowski et al., 2008), iterative mapping (Manske and Kwiatkowski, 2009), read realignment (Albers et al., 2010; Homer and Nelson, 2010; Li, 2011; Depristo et al., 2011) and local assembly (Carnevali et al., 2011). <strong>However, because these methods still rely on the initial mapping, it is difficult for them to identify and recover mismapped or unmapped reads due to high sequence divergence, long insertions, SVs, copy number changes or misassemblies of the reference genome. They have not solved the problem from the root.</strong>"
        <footer> - <a href='http://bioinformatics.oxfordjournals.org/content/early/2012/05/07/bioinformatics.bts280.abstract'>Heng Li, 2012</a></footer>
        </blockquote>
        </li>
        </ul>
        <h7><a href="#TOC">Table of contents</a></h7>
        </div>

        <div id='ARCInstallation'>
        <h3>Installation</h3>
        <h4>Prerequisites:</h4>

        A mapper and assembler must be installed in your path, the following are currently supported.
        <ul>
        <li>
          Mapper:
          <ul>
            <li> <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie 2 </a></li>
            <li> <a href="http://genome.ucsc.edu/FAQ/FAQblat.html#blat3">Blat</a>
            <p>Note that Blat currently only supports FASTA format as an input for query sequences. We have contacted the author with a patch to add FASTQ support and hope that it will be incorporated in the release version soon. Until then we supply a patch which the user must apply in order to add FASTQ support to BLAT. </p>
            </li>
          </ul>
        <li>
          Assembler:
            <ul>
              <li><a href="http://www.454.com/products/analysis-software/">Roche/Newbler assembler</a></li>
              <li><a href="http://bioinf.spbau.ru/spades/">Spades assembler</a></li>
            </ul>
        </li>
          For more information on installing these programs see the <a href="https://github.com/ibest/ARC/wiki/Details-on-installation">Advanced Installation Instructions</a> on the <a href="https://github.com/ibest/ARC/wiki/">Wiki.</a> Any combination of the mappers and assemblers above should work equally well in most cases, however there can be differences in recruited reads, sensitivity to repeats, and assembly results, so testing different combinations may improve results.

        </li>
        </ul>
        <h4>Additional Requirements:</h4>
        <ul>
          <li><a href="http://www.python.org/getit/">Python 2.7.X</a></li>
          <li><a href="http://biopython.org/wiki/Download">Python module BioPython </a> (developed with v 1.6.0, but should work with any recent release)</li>

        </ul>


      <h4>Download ARC:</h4>
      <h5>Option 1: Using "git clone"</h5>
        <pre><code>$ git clone git://github.com/ibest/ARC.git</code></pre>
      <section id="downloads2" class="clearfix">
      <h5>Option 2: Compressed archive</h5>
        <a href="https://github.com/ibest/ARC/zipball/master" id="download-zip" class="button"><span>Download .zip</span></a>
        <a href="https://github.com/ibest/ARC/tarball/master" id="download-tar-gz" class="button"><span>Download .tar.gz</span></a>
      </section>
      <br>

      <h4>ARC installation options:</h4>
      ARC can be set up in a number of ways depending on your use case.

      <h5>Option 1: Run ARC without installing</h5>
      <ul><li>
      ARC can be run from the "git clone" or uncompressed archive folder without installing:
      <pre><code> $ ./ARC/bin/ARC > log.txt</code></pre>
      Note that while it isn't necessary to pipe the output from ARC to a log file, it is recommended in case the need to troubleshoot arises.
      </li></ul>

      <h5>Option 2: Install as a Python module</h5>
      <ul><li>
      ARC supports the standardized Python <a href="https://docs.python.org/2/install/">Distutils</a> approach to installing modules. Run the following to install to your default Python path, and see the <a href="https://docs.python.org/2/install/">Distutils documentation</a> for more advanced options:
      <pre><code> $ python setup.py install</code></pre>
      </li></ul>

      <h5>Option 3: Install using a <a href="http://docs.python-guide.org/en/latest/dev/virtualenvs/">Python Virtual Environment</a></h5>
      <ul><li>
        Move to the directory where you keep all of your python virtual environments and run the following commands:
<pre><code>$ cd ~/pyenvs
$ virtualenv arc
$ source /path/to/ARC/bin/activate
$ cd /path/to/arc/source
$ python setup.py install
</code></pre>
      </li></ul>
      <h7><a href="#TOC">Table of contents</a></h7>
    </div>

<div id="PatchingBLAT">
    <h3> Patching BLAT to support FASTQ (Optional):</h3>
    We have found that the BLAT mapper by Jim Kent is an excellent, fast, and efficient tool for aligning reads against a divergent reference. Unfortunately it doesn't natively support the FASTQ format which is the standard output of most HTS platforms. Ilya Zhbannikov (author of <a href="https://bitbucket.org/izhbannikov/seqyclean">SeqyClean</a>) wrote a patch which adds FASTQ support to BLAT. This patch is distributed in the "contrib" folder. Please ensure that you have the proper academic/nonprofit status, or acquire a license from <a href="(http://www.kentinformatics.com/">Kent Informatics</a>) and then use the following set of commands to download, patch, and install BLAT with FASTQ support.
<pre><code>$ wget http://users.soe.ucsc.edu/~kent/src/blatSrc.zip
$ unzip blatSrc.zip
$ patch -p0 &lt/path/to/ARC/contrib/blat+fastq_support.patch
$ cd blatSrc
$ export MACHTYPE=x86_64
$ mkdir ~/bin
$ mkdir ~/bin/x86_64
$ make
</code></pre>
  Executables for blat with FASTQ support will now be located in your ~/bin/x86_64 folder. Either add this folder to your path, or move the executable files to a folder already in your path.
<br><h7><a href="#TOC">Table of contents</a></h7>
    </div>

<div id='ARCTesting'>
    <h3>Testing your ARC install:</h3>
      A small test dataset and configuration file is included with ARC in the "test_data" folder. This can be used to test ARC, either by running the included shell script, or by running ARC normally: <br>
<pre><code>$ cd test_data
$ ./runarc
</code></pre>
Or (if ARC is installed in the PATH)
<pre><code>$ cd test_data
$ ARC
</code></pre>
<h7><a href="#TOC">Table of contents</a></h7>
    </div>


<div id="ARCUsage">
<h3>Usage</h3>
<ul>
  <li>First, configure input files.
    <ul>
      <li>Combine reads into a maximum of 3 files per sample: PE1, PE2, SE (where PE stands for Paired End, and SE for Single End). These files can be fasta or fastq formatted but must all be the same format.</li>
      <li>Ensure that the targets file is in fasta format and that all entries have unique names.</li>
      <li>Create a file named ARC_config.txt (see the files in test_data for an example). Put this file in a working directory on a drive with plenty of free space.</li>
    </ul></li>
    <li>Run ARC using the approach appropriate to the installation method selected.
      <pre><code>$ ARC > log.txt </code></pre>
    </li>
    </ul>
    <h7><a href="#TOC">Table of contents</a></h7>
</div>

<div id="ARCOutputs">
<h3>Outputs</h3>

ARC will create a set of folders corresponding to the samples included in the ARC_config.txt. These will be labeled working_* and finished_*.<br>

The working_* folders contain the following files:
<ul>
    <li>I00n_contigs.fasta - these are the intermediate contig results for each iteration
    mapping_log.txt - the output produced by the mapper</li>
    <li>*.idx files - Biopython indexes of the reads (these are kept to facilitate restarting without re-indexing the read files)</li>
</ul>
The finished_* folders contain the following files:
<ul>
    <li>contigs.fasta - this contains all contigs for the sample
        Contig names have the following format: SampleID_:TargetID:_ContigN where there can be multiple contigs for one target</li>
    <li>PE1.fastq, PE2.fastq, SE.fastq - these files contain the reads which were mapped on the final iteration of ARC.
        Reads have a slightly modified name to ensure compatibility with Newbler and contain a Sample_:_Target field in the description.</li>
</ul>
<h7><a href="#TOC">Table of contents</a></h7>
</div>

<div id="ARCConfig">
<h3>Configuring ARC</h3>
<h4>Command line options include:</h4>
<p>
<table>
<tr>
  <th>Option</th>
  <th>Description</th>
</tr>
  <tr><td>-d</td><td>--debug, Turn on extra debug output.</td></tr>
  <tr><td>-p</td><td>--profile, Turn on code profiling.</td></tr>
  <tr><td>-c</td><td>--config, Specify an alternate ARC config file (ARC_config.txt by default).</td></tr>
  <tr><td>-v</td><td>--version, Print the version number and exit.</td></tr>
</table>
</p>
All configuration for ARC is stored in the ARC_config.txt file. Consult the example configuration file in test_data for exact formatting requirements. The following options are currently supported:
<p>
<table>
<tr>
  <th>Parameter</th><th>Description</th>
</tr>




<tr><td>reference*</td><td>A fasta file contain one or more reference sequences.</td></tr>
<tr><td>numcycles</td><td>Maximum number of mapping and assembly cycles ARC will carry out Default: 1</td></tr>
<tr><td>max_incorporation</td><td>Control for repeat elements. If total reads recruited in the current cycle is greater than max_incorporation X reads recruited in previous cycle, assembly will not be carried out. Default: 5</td></tr>
<tr><td>bowtie2_k</td><td>Controls the max number of matches Bowtie 2 will report for each read. Default 5</td></tr>
<tr><td>format*</td><td>Format for files containing reads, can be fasta or fastq.</td></tr>
<tr><td>mapper*</td><td>Mapper to use during read recruitment, can be bowtie2 or blat.</td></tr>
<tr><td>assembler*</td><td>Assembler to use during assembly stage, can be newbler or spades</td></tr>
<tr><td>urt</td><td>Newbler parameter “use read tips” may reduce the number of ARC iterations by instructing Newbler to extend contigs using single reads at the edges of contigs. Note that ARC will not use 'urt' on the final iteration to ensure higher quality contigs. Default False</td></tr>
<tr><td>verbose</td><td>Output extensive logging details about ARC operation including all calls to external programs Default False</td></tr>
<tr><td>assemblytimeout</td><td>Amount of time (in minutes) ARC will wait for an assembly to finish before killing the assembly process. Adjusting this value can make assemblies of large targets possible, or reduce the impact of repeats on large ARC runs. Default 10.</td></tr>
<tr><td>cdna</td><td>Newbler parameter that enables experimental RNAseq assembly and read incorporation reporting. Newbler will be run in transcriptome assembly mode on the final ARC iteration. Default: False</td></tr>
<tr><td>rip</td><td>Newbler parameter that instructs Newbler to only place reads in a single contig. In some cases Newbler will split a read placing parts of it in more than one contig. Default: False</td></tr>
<tr><td>subsample</td><td>Subsample read depth to a percentage of the original number of mapped reads. In cases where sequencing depth is great (>100x) it is often beneficial to only assemble a random subset of the mapped reads. For example, subsample=0.4 would cause ARC to retain 40% of mapped reads for assembly. Default: 1</td></tr>
<tr><td>maskrepeats</td><td>Causes ARC to mask simple tandem repeats in contigs before mapping. This results in recruitment of fewer reads contain repeats. Default: True</td></tr>
<tr><td>nprocs</td><td>Number of processors ARC should use. ARC can effectively make use of at least 64 cores when processing large jobs. Default: 1</td></tr>
<tr><td>fastmap</td><td>BLAT mapper parameter, runs BLAT in fastMap mode that requires high identity and doesn't allow insertions or deletions.</td></tr>
<tr><td>sloppymapping</td><td>Controls whether reads will be mapped in low-specificity (sloppy) mode on the first iteration. Default True.</td></tr>
</table>
</p>
<h7><a href="#TOC">Table of contents</a></h7>
</div>



<div id="TipsAndTricks">
<h3>Tips and tricks for advanced users</h3>
<h4>Using multiple sequences for a single target:</h4>
  In some cases it may be difficult to identify a single best homologous target to use in seeding an assembly. In this situation ARC can be configured to use multiple sequences as a single target by using the following naming scheme:
      <ul>
        <li>Name the targets sequences so that they share a common middle element in the following way:
          <ul>
            <li><strong>></strong>ID_:Target1:_seq1</li>
            <li><strong>></strong>ID_:Target1:_seq2</li>
            </ul>
        </li>
      </ul>

<h4>Setting map_against_reads:</h4>
  Only a small number of reads may map on the first iteration if your reference is very distantly related or coverage is low. In these situations, set map_against_reads=True in the ARC_config.txt file. This will use all of the mapped reads as the new set of targets instead of using contigs on the first iteration. Please note that this feature is experimental and sometimes leads to incorporation of repetitive elements and off-target contigs.
  <br>
<h7><a href="#TOC">Table of contents</a></h7>
</div>


<div id="Contact">
<h3>Questions/Comments/Suggestions</h3>
Please post questions, comments and suggestions to the <a href="https://groups.google.com/forum/#!forum/arc-assembly">ARC user group</a>, or email (arc-assembly {at} googlegroups {dot} com).
</div>


<script type="text/javascript">
  var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
  document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
    try {
      var pageTracker = _gat._getTracker("UA-39120406-2");
    pageTracker._trackPageview();
    } catch(err) {}
</script>

      </div>
    </div>
  </body>
</html>



<!--

<br><br>
<pre><code>$ cd your_repo_root/repo_name
$ git fetch origin
$ git checkout gh-pages
</code></pre>

<p>If you're using the GitHub for Mac, simply sync your repository and you'll see the new branch.</p>

<h3>
<a name="designer-templates" class="anchor" href="#designer-templates"><span class="octicon octicon-link"></span></a>Designer Templates</h3>

<p>We've crafted some handsome templates for you to use. Go ahead and continue to layouts to browse through them. You can easily go back to edit your page before publishing. After publishing your page, you can revisit the page generator and switch to another theme. Your Page content will be preserved if it remained markdown format.</p>

<h3>
<a name="rather-drive-stick" class="anchor" href="#rather-drive-stick"><span class="octicon octicon-link"></span></a>Rather Drive Stick?</h3>

<p>If you prefer to not use the automatic generator, push a branch named <code>gh-pages</code> to your repository to create a page manually. In addition to supporting regular HTML content, GitHub Pages support Jekyll, a simple, blog aware static site generator written by our own Tom Preston-Werner. Jekyll makes it easy to create site-wide headers and footers without having to copy them across every page. It also offers intelligent blog support and other advanced templating features.</p>

<h3>
<a name="authors-and-contributors" class="anchor" href="#authors-and-contributors"><span class="octicon octicon-link"></span></a>Authors and Contributors</h3>

<p>You can <a href="https://github.com/blog/821" class="user-mention">@mention</a> a GitHub username to generate a link to their profile. The resulting <code>&lt;a&gt;</code> element will link to the contributor's GitHub Profile. For example: In 2007, Chris Wanstrath (<a href="https://github.com/defunkt" class="user-mention">@defunkt</a>), PJ Hyett (<a href="https://github.com/pjhyett" class="user-mention">@pjhyett</a>), and Tom Preston-Werner (<a href="https://github.com/mojombo" class="user-mention">@mojombo</a>) founded GitHub.</p>

<h3>
<a name="support-or-contact" class="anchor" href="#support-or-contact"><span class="octicon octicon-link"></span></a>Support or Contact</h3>

<p>Having trouble with Pages? Check out the documentation at <a href="http://help.github.com/pages">http://help.github.com/pages</a> or contact <a href="mailto:support@github.com">support@github.com</a> and we’ll help you sort it out.</p>
        </section>

        <footer>
          Arc is maintained by <a href="https://github.com/ibest">ibest</a><br>
          This page was generated by <a href="http://pages.github.com">GitHub Pages</a>. Tactile theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.
        </footer>

-->
